src/main/java/seedu/address/logic/commands/OpenCommand.java

```diff
@@ -44,29 +66,95 @@ public OpenCommand(TagName tagName) {
      */
     @Override
     public CommandResult execute(Model model) throws CommandException {
-        // Check if tagName is in tag list
-        List<Tag> tagList = model.findFilteredTagList(new TagNameEqualsKeywordPredicate(tagName));
-        if (tagList.isEmpty()) {
-            throw new CommandException(String.format(MESSAGE_TAG_NOT_FOUND, tagName.toString()));
+        assert label != null || tagName != null;
+
+        List<Tag> tagsToOpen = new ArrayList<>();
+
+        if (tagName != null) {
+            // Open tag based on tag name
+            // Check if tagName is in tag list
+            List<Tag> tagList = model.findFilteredTagList(new TagNameEqualsKeywordPredicate(tagName));
+            if (tagList.isEmpty()) {
+                throw new CommandException(String.format(MESSAGE_TAG_NOT_FOUND, tagName.toString()));
+            }
+            assert tagList.size() == 1 : "Only one tag should be found.";
+            tagsToOpen = List.of(tagList.get(0));
+        } else {
+            // Open tag based on label
+            List<Tag> tagList = model.findFilteredTagList(new TagLabelEqualsKeywordPredicate(label));
+            if (tagList.isEmpty()) {
+                throw new CommandException(String.format(MESSAGE_LABEL_NOT_FOUND, label.toString()));
+            }
+            tagsToOpen = tagList;
         }
 
-        // Get tag and prepare file to be opened
-        Tag tagToOpen = tagList.get(0);
-        File file = new File(tagToOpen.getFileAddress().value);
+        // Perform open and get a set of opened tags
+        Map<Tag, String> openResults = openTags(tagsToOpen);
 
-        try {
-            // Open file using java.awt.Desktop
-            Desktop.getDesktop().open(file);
-        } catch (IOException | IllegalArgumentException e) {
-            throw new CommandException(e.getMessage(), e);
+        // Format result string
+        StringBuilder sb = new StringBuilder();
+        boolean hasError = false;
+        for (Tag tag : tagsToOpen) {
+            String message = openResults.get(tag);
+            if (message.startsWith("Error")) {
+                hasError = true;
+            }
+            sb.append(message).append("\n");
         }
-        return new CommandResult(String.format(MESSAGE_SUCCESS, tagToOpen.toString()));
+
+        String resultMessage = sb.toString();
+        // Remove last "\n"
+        if (resultMessage.endsWith("\n")) {
+            resultMessage = resultMessage.substring(0, resultMessage.length() - 1);
+        }
+
+        // Throw exception if some tags were not opened
+        if (hasError) {
```

Comment:
Does this exception ever gets thrown? Would the openTags() method have already handled the tag not opened exception?

Reply:
It's thrown afterwards, so that all tags that can be opened are opened.